README
Cheferize : A Ruby library to 'translate' strings into Swedish Chef.
Copyright 2010 by Alex Dean.

Usage:
>> class String; include Cheferize; end 
>> "This is a test!".to_chef
=> "Tees is e test, bork bork bork!"

Cheffereeze-a is free-a sufftvere; yuoo cun redeestriboote-a it und/ur mudeeffy 
it under zee terms ooff zee GNU Generel Poobleec Leecense-a es poobleeshed by 
zee Free-a Sufftvere-a Fuoondeshun; ieezeer ferseeun 2 ooff zee Leecense, oor 
(et yuoor oopshun) uny leter ferseeun. 

Cheffereeze-a is deestribooted in zee hupe-a tet it veell be-a useffool, 
boot VITOUT UnY VERRUnTY; veetuoot ifee zee impleeed verrunty ooff 
MERCHUnTEBILITY oor FITNESS FOR E PERTICULER PURPOSE. See-a zee 
GNU Generel Poobleec Leecense-a fur mure-a deteeels. 

Yuoo shuoold hefe-a receeefed e cupy ooff zee GNU Generel Poobleec Leecense-a 
elung veet Cheffereeze; iff nut, vreete-a tu zee Free-a Sufftvere-a 
Fuoondeshun, Inc., 51 Frunkleen St, Feefft Fluur, Bustun, MA 02110-1301 USA

Inspired by encheferizer.php by eamelink (Erik Bakker).
About encheferizer.php :
  http://bork.eamelink.nl (defunct as of 17 Jan 2010)
  Based on the original chef.x from John Hagerman.
  More info about the original chef.x : http://tbrowne.best.vwh.net/chef/ (also defunct as of 17 Jan 2010)


CHANGELOG.md
# CHANGELOG

## 0.1.0 - August 13, 2019

First release as a gem. (Project was just a bare ruby script prior to this.)

## 0.0.0 - January, 2010

Initial creation of this project. See commit log for dates.


test/translate.php
#!/usr/bin/php
<?
  require_once '../encheferizer.php';
  $e = new encheferizer();
  
  $source = fopen( "american-english", "r" );
  $target = fopen( "php-out", "w" );
  
  while ( ! feof( $source ) ) {
      $buffer = fgets( $source, 4096 );
      fwrite( $target, $e->encheferize( $buffer ) );
  }
  fclose( $source );
  fclose( $target );
?>

spec/lib/cheferize_spec.rb
require 'spec_helper'
require_relative '../../lib/cheferize'

class String
  include Cheferize
end

RSpec.describe Cheferize do
  describe '#to_chef' do
    it 'does the bork' do
      subject = 'hello world'
      expect(subject.to_chef).to eq 'hellu vurld'
    end

    it 'does more borking' do
      subject = 'the quick brown fox jumped over the lazy dog'
      expect(subject.to_chef).to eq 'zee qooeeck broon fux joomped oofer zee lezy dug'
    end
  end
end


spec/lib/cheferize/chef_string_spec.rb
require 'spec_helper'
require_relative '../../../lib/cheferize/chef_string'

RSpec.describe Cheferize::ChefString do
  describe '#in_same_case_as' do
    it 'returns target string transformed into the same case as the first character of target' do
      subject = Cheferize::ChefString.new('a')
      expect(subject.in_same_case_as('B')).to eq 'A'

      subject = Cheferize::ChefString.new('a')
      expect(subject.in_same_case_as('b')).to eq 'a'

      subject = Cheferize::ChefString.new('A')
      expect(subject.in_same_case_as('B')).to eq 'A'

      subject = Cheferize::ChefString.new('A')
      expect(subject.in_same_case_as('b')).to eq 'a'
    end
  end
end


encheferizer.rb
# Usage:
# Cheferize can be included into any class which responds to to_s.  
# It adds a method to_chef.
# class String
#   include Cheferize
# end
# 
# >> "This is a test!".to_chef
# => "Tees is e test!"
#
# These are applied sequentially to each character in the source string until 1 matches.
module Cheferize
  
  def to_chef
    # split string into words
    # TODO : punctuation will break rules which assert :last position.
    output = to_s.split( ' ' ).map do |word|
      cheferize_word( word )
    end
    output.join( ' ' )
  end
  
  def cheferize_word( input )
    
    i=0
    output = ''
    # see rule "interior ir becomes ur, or first-occurring interior i becomes ee"
    @chef_saw_an_i = nil
    
    while i < input.size do
      # a single character will be :first but not :last.
      position = nil
      if i == 0
        position = :first
      elsif i == ( input.size - 1 )
        position = :last
      end
      
      # apply rules to each character, breaking on the first one which returns output
      # for the word 'testify', the rules will see 'testify', 'estify', 'stify', 'tify', etc...
      subject = input[ i..(input.size-1) ]
      
      Cheferize.rules.each do |rule|
        out = self.send( rule, subject, position )
        # uncomment to see to_chef in action.
        #puts "#{subject} #{' ('+position.to_s+') ' if position}: '#{rule.to_s}' returns '#{out.inspect}'"
        if out
          output += out[ 0 ]
          i += out[ 1 ]
          break
        end
      end
    end
    
    output
  end
  
  # General rules
  # after a replacement occurs, skip to the next source character.  (only 1 transformation per letter)
  # if a multi-char test matches, skip over to the next chars. 'few' matches 'ew->oo'.  Then we're done.
  
  # for first char
    # ^e->i
    # ^o->oo
  # for non-first chars
    # ew->oo
    # e$->e-a
    # f->ff
    # ir->ur, or i->ee if no i has occurred.
    # ow->oo or o->u
    # tion -> shun
    # u->oo

  # An->Un
  # Au->Oo
  # A.->E (A not at the end becomes E)
  # en$->ee
  # e not at the beginning, do nothing. Pointless?
  # th->tt
  # the->zee
  # v->f
  # w->v
  # pass
  
  # Define a transformation rule.
  # Block must accept subject & position parameters.
  def self.rule( name, &block )
    @rules ||= []
    sym = name.to_sym
    @rules << sym
    define_method( sym, block )
  end

  # Shows the transliteration rules.
  def self.rules
    @rules
  end
  
  # no recursive cheferizing yet.
  rule "pass the bork" do |subject, position|
    [ subject, 4 ] if position == :first && subject[ 0,4 ].downcase == 'bork'
  end
  
  rule "randomly borkify after . or !" do |subject, position|
    if position == :last && ( subject[ 0,1 ] == '.' || subject[ 0,1 ] == '!' ) && rand( 3 ) == 0
      [ ', bork bork bork!', 1 ]
    end
  end
  
  # ^e->i
  rule "initial e becomes i" do |subject, position|
    if position == :first && subject[ 0,1 ].downcase == 'e'
      [ ChefString.new( 'i' ).in_same_case_as( subject[ 0,1 ] ), 1 ]
    end
  end
  
  # ^o->oo
  rule "initial o becomes oo" do |subject, position|
    if position == :first && subject[ 0,1 ].downcase == 'o'
      [ ChefString.new( 'o' ).in_same_case_as( subject[ 0,1 ] ) + 'o', 1 ]
    end
  end
  
  # ew->oo, if not at the beginning
  rule "interior ew becomes oo" do |subject, position|
    [ 'oo', 2 ] if position != :first && subject[ 0,2 ] == 'ew'
  end
  
  # e$->e-a, if at the end
  rule "final e becomes e-a" do |subject, position|
    [ 'e-a', 1 ] if position == :last && subject[ 0,1 ] == 'e'
  end
  
  # f->ff if not at the beginning
  rule "interior f becomes ff" do |subject, position|
    [ 'ff', 1 ] if position != :first && subject[ 0,1 ] == 'f'
  end
  
  # ir->ur, or i->ee if no i has occurred.
  rule "interior ir becomes ur, or first-occurring interior i becomes ee" do |subject, position|
    if position != :first
      if subject[ 0,2 ] == 'ir'
        [ 'ur', 2 ]
      elsif subject[ 0,1 ] == 'i' && ! @chef_saw_an_i
        @chef_saw_an_i = true
        [ 'ee', 1 ]
      end
    end
  end
  
  # ow->oo or o->u
  rule "ow becomes oo, or o becomes u" do |subject, position|
    if position != :first
      if subject[ 0,2 ] == 'ow'
        [ 'oo', 2 ]
      elsif subject[ 0,1 ] == 'o'
        [ 'u', 1 ]
      end
    end
  end
  
  # tion -> shun
  rule "tion becomes shun" do |subject, position|
    if position != :first && subject[ 0,4 ] == 'tion'
      [ 'shun', 4 ]
    end
  end
  
  # u->oo
  rule "interior u becomes oo" do |subject, position|
    if position != :first && subject[ 0,1 ] == 'u'
      [ 'oo', 1 ]
    end
  end
  
  # An->Un, an->un
  rule "an becomes un" do |subject, position|
    if subject[ 0,2 ].downcase == 'an'
      [ ChefString.new( 'u' ).in_same_case_as( subject[ 0,1 ] ) + 'n', 2 ]
    end
  end
  
  # Au->Oo, au->oo
  rule "au becomes oo" do |subject, position|
    if subject[ 0,2 ].downcase == 'au'
      [ ChefString.new( 'o' ).in_same_case_as( subject[ 0,1 ] ) + 'o', 2 ]
    end
  end
  
  # A.->E (A not at the end becomes E)
  rule "non-final a becomes e" do |subject, position|
    if position != :last && subject[ 0,1 ].downcase == 'a'
      [ ChefString.new( 'e' ).in_same_case_as( subject[ 0,1 ] ), 1 ]
    end
  end
  
  # en$->ee
  rule "final en becomes ee" do |subject, position|
    if position != :first && subject == 'en'
      [ 'ee', 2 ]
    end
  end
  
  # e not at the beginning, do nothing. Pointless?
  
  # the->zee
  rule "the becomes zee" do |subject, position|
    if subject[ 0,3 ].downcase == 'the'
      [ ChefString.new( 'z' ).in_same_case_as( subject[ 0,1 ] ) + 'ee', 3 ]
    end
  end
  
  # th->t
  rule "th becomes t" do |subject, position|
    if subject[ 0,2 ].downcase == 'th'
      [ ChefString.new( 't' ).in_same_case_as( subject[ 0,1 ] ), 2 ]
    end
  end
  
  # v->f
  rule "v becomes f" do |subject, position|
    if subject[ 0,1 ].downcase == 'v'
      [ ChefString.new( 'f' ).in_same_case_as( subject[ 0,1 ] ), 1 ]
    end
  end
  
  # w->v
  rule "w becomes v" do |subject, position|
    if subject[ 0,1 ].downcase == 'w'
      [ ChefString.new( 'v' ).in_same_case_as( subject[ 0,1 ] ), 1 ]
    end
  end
  
  rule "pass on anything else" do |subject, position|
    [ subject[ 0,1 ], 1 ]
  end
  
  # A ChefString provides an additional public method for String.  in_same_case_as.
  class ChefString
    instance_methods.each { |m| undef_method m unless m =~ /(^__|^send$|^object_id$)/ }

    def initialize( str )
      @target = str
    end

    # return target string, transformed into the same case as the first character of target.
    def in_same_case_as( target )
      to_case( @target, get_case( target ) )
    end

    def get_case( letter )
      code = letter[ 0 ]
      if code > 96 && code < 123
        :lower
      elsif code > 64 && code < 91
        :upper
      else
        :non_alpha
      end
    end

    def to_case( letter, sym )
      case sym
      when :upper
        letter.upcase
      when :lower
        letter.downcase
      else
        letter
      end
    end

    protected

    # proxy any remaining calls to the underlying string.
    def method_missing(name, *args, &block)
      @target.send(name, *args, &block)
    end
  end
  
end

encheferizer.php
<?php
#########################################################
#                                                       #
#  PHP Encheferizer 0.9 (20-5-2003)                     #
#  Convert English to Mock Swedish.                     #
#  Ported to PHP by eamelink (Erik Bakker)              #
#  PHP Encheferizer website : http://bork.eamelink.nl   #
#                                                       #     
#  Based on the original chef.x from John Hagerman.     #
#  More info about the original chef.x :                #
#  http://tbrowne.best.vwh.net/chef/                    #
#                                                       #
#  Works like :                                         #
#    $chef = new encheferizer();                        #
#    $swedish = $chef->encheferize($string);            #
#                                                       #
#########################################################
class encheferizer
{
    function encheferizer() {}

    function encheferize($string)
    {
        $delimiters = " \n\t\\,<.>/?;:'\"[{]}|=+-_!@#$%^&*()~`";
        $newline = "";
        $word = "";
        
        for($i = 0; $i < strlen($string); $i++){
            $s = substr($string, $i, 1);
            
            // if this character is not a delimiter, accumulate it into $word
            if(($position = strstr($delimiters, $s)) === false){ 

                $word .= $s;
  
            // when a delimiter is found
             } else {
               // encheferize a non-empty $word and add to $newline
                if(!empty($word)){
                    $newline .= $this->encheferizeWord($word);
                    $word = "";
                }
                // if . or !, randomly borkify
                if($s == "." || $s == "!"){
                    if(rand(0,3) == 1){  
                        $newline .= ", Bork Bork Bork!";
                        
                    } else {
                        $newline .= ".";
                    }
                // pass other delimiters through unchanged
                } else {
                    $newline .= $s;
                }
            }
        }
        $newline .= $this->encheferizeWord($word);
        return $newline;    
    }
    
    function encheferizeWord($word)
    {
        if(strtolower($word) == "bork"){
            return $word;
        }
        $newword = "";    

        $i = 0;
        while($i < strlen($word)){
            $char = substr($word, $i, 1);
            $isLast = ($i == strlen($word) - 1) ? true : false;   
            $iSeen = false;
            
            // rules for these letters beginning a word
            if($i == 0){ 
                switch($char){
                case "e":
                    $newword = "i";
                    $i++;
                    continue 2;
                    break;
                case "E":
                    $newword = "I";
                    $i++;
                    continue 2;
                    break;
                case "o":
                    $newword = "oo";
                    $i++;
                    continue 2;
                    break;
                case "O":
                    $newword = "Oo";
                    $i++;
                    continue 2;
                    break;
                }
            
            // rules for these letters not beginning a word
            } else {
                // ew->oo
                // e$->e-a
                if($char == "e"){
                    if(!$isLast && substr($word, $i + 1, 1) == "w"){
                        $newword .= 'oo';
                        
                        $i+=2; 
                        continue;
                    } elseif($isLast){
                        $newword .= 'e-a';
                        $i++;
                        continue;
                    }
                // f->ff
                } elseif($char == "f"){
                    $newword .= "ff"; 
                    $i++;
                    continue;
                // ir->ur or i->ee if no i has occurred.
                } elseif($char == "i"){
                    
                    if(!$isLast && substr($word, $i + 1, 1) == "r"){
                        $newword .= "ur";
                        $i+=2;
                        continue;
                    } elseif(!$iSeen){
                        $newword .= "ee";
                        $i++;
                        
                        $iSeen = true;
                        continue;
                    }
                // ow->oo or o->u
                } elseif($char == "o"){
                    if(!$isLast && substr($word, $i + 1, 1) == "w"){
                        $newword .= "oo";
                        $i+=2;
                        continue;
                    } else {
                        $newword .= "u";
                        $i++;
                        continue;
                    }
                // tion -> shun
                } elseif($i <= strlen($word - 4) && $char == "t" && substr($word, $i + 1, 1) == "i" 
                            && substr($word, $i + 2, 1) == "o" && substr($word, $i + 3, 1) == "n"){  // Caption -> Capshun
                    $newword .= "shun";
                    $i += 4;
                    continue;
                // u->oo
                } elseif($char == "u"){
                    $newword .= "oo";
                    $i++;
                    continue;
                } elseif($char == "U"){
                    $newword .= "Oo";
                    $i++;
                    continue;
                }
            } // En of In-word rules
            
            
            // Things that may be replaced anywhere
            
            if($char == "A"){
                // An->Un
                if(!$isLast && substr($word, $i+1, 1) == "n"){
                    $newword .= "Un";
                    $i+=2;
                    continue;
                // Au->Oo
                } elseif(!$isLast && substr($word, $i+1, 1) == "u"){
                    $newword .= "Oo";
                    $i+=2;
                    continue;
                // A.->E
                } elseif(!$isLast){
                    $newword .= "E";
                    $i++;
                    continue;
                } 
            } elseif($char == "a"){
                if(!$isLast && substr($word, $i+1, 1) == "n"){
                    $newword .= "un";
                    $i+=2;
                    continue;
                } elseif(!$isLast && substr($word, $i+1, 1) == "u"){
                    $newword .= "oo";
                    $i+=2;
                    continue;
                } elseif(!$isLast){
                    $newword .= "e";
                    $i++;
                    continue;
                } 
            
            
            
            } elseif($char == "e"){
                // en$->ee
                if(!$isLast && substr($word, $i+1, 1) == "n" && $i == strlen($word) - 2){
                    $newword .= "ee";
                    $i+=2;
                    continue;
                // can this ever occur?
                } elseif ($i > 0){
                    // Do nothing
                    
                }
            } elseif($char == "t"){
              
                // th->t
                if($i == strlen($word) -2 && substr($word, $i+1, $i) == "h"){
                    $newword .= "t";
                    $i+=2;
                    continue;
                // the->zee
                } elseif($i <= strlen($word) - 3 && substr($word, $i+1, 1) == "h" && substr($word, $i+2, 1) == "e"){
                    $newword .= "zee";
                    $i+=3;
                    continue;
                }
            } elseif($char == "T" && $i<= strlen($word)-3 && substr($word, $i+1, 1) == "h"
                        && substr($word, $i+2, 1) == "e"){
                $newword .= "Zee";
                $i +=3;
                continue;
            // v->f
            } elseif($char == "v"){
                $newword .= "f";
                $i++;
                continue;
            } elseif($char == "V"){
                $newword .= "F";
                $i++;
                continue;
            // w->v
            } elseif($char == "w"){
                $newword .= "v";
                $i++;
                continue;
            } elseif($char == "W"){
                $newword .= "V";
                $i++;
                continue;
            } 
            
            $newword .= $char;
            $i++;   
                
                
                
        } // End of FOR lus
        
        return $newword;
    }
}
?>   

lib/cheferize/chef_string.rb
module Cheferize
  # A ChefString provides an additional public method for String.  in_same_case_as.
  class ChefString
    instance_methods.each { |m| undef_method m unless m =~ /(^__|^send$|^object_id$)/ }

    def initialize(str)
      @target = str
    end

    # return target string, transformed into the same case as the first character of target.
    def in_same_case_as(target)
      to_case(@target, get_case(target))
    end

    def get_case(letter)
      code = letter.codepoints[0]
      if code > 96 && code < 123
        :lower
      elsif code > 64 && code < 91
        :upper
      else
        :non_alpha
      end
    end

    def to_case(letter, sym)
      case sym
      when :upper
        letter.upcase
      when :lower
        letter.downcase
      else
        letter
      end
    end

    protected

    # proxy any remaining calls to the underlying string.
    def method_missing(name, *args, &block)
      @target.send(name, *args, &block)
    end
  end
end


lib/cheferize.rb
require "cheferize/version"
require "cheferize/chef_string"

# Cheferize : A Ruby library to 'translate' strings into Swedish Chef.
# Copyright 2010 by Alex Dean <github@mostlyalex.com>.
#
# Usage:
# >> class String; include Cheferize; end
# >> "This is a test!".to_chef
# => "Tees is e test, bork bork bork!"
#
# Cheffereeze-a is free-a sufftvere; yuoo cun redeestriboote-a it und/ur mudeeffy
# it under zee terms ooff zee GNU Generel Poobleec Leecense-a es poobleeshed by
# zee Free-a Sufftvere-a Fuoondeshun; ieezeer ferseeun 2 ooff zee Leecense, oor
# (et yuoor oopshun) uny leter ferseeun.
#
# Cheffereeze-a is deestribooted in zee hupe-a tet it veell be-a useffool,
# boot VITOUT UnY VERRUnTY; veetuoot ifee zee impleeed verrunty ooff
# MERCHUnTEBILITY oor FITNESS FOR E PERTICULER PURPOSE. See-a zee
# GNU Generel Poobleec Leecense-a fur mure-a deteeels.
#
# Yuoo shuoold hefe-a receeefed e cupy ooff zee GNU Generel Poobleec Leecense-a
# elung veet Cheffereeze; iff nut, vreete-a tu zee Free-a Sufftvere-a
# Fuoondeshun, Inc., 51 Frunkleen St, Feefft Fluur, Bustun, MA 02110-1301 USA
#
# Inspired by encheferizer.php by eamelink (Erik Bakker).
# About encheferizer.php :
#   http://bork.eamelink.nl (defunct as of 17 Jan 2010)
#   Based on the original chef.x from John Hagerman.
#   More info about the original chef.x : http://tbrowne.best.vwh.net/chef/ (also defunct as of 17 Jan 2010)
#
module Cheferize
  def self.included(other)
    if !other.respond_to?(:to_s)
      raise "#{other} cannot include Cheferize.  #{other} does not respond to to_s."
    end
  end

  def to_chef
    # TODO : punctuation will break rules which assert :last position.
    output = to_s.split(' ').map do |word|
      cheferize_word(word)
    end
    output.join(' ')
  end

  def cheferize_word(input)
    i=0
    output = ''
    # see rule "interior ir becomes ur, or first-occurring interior i becomes ee"
    @chef_saw_an_i = nil

    while i < input.size do
      # a single character will be :first but not :last.
      position = nil
      if i == 0
        position = :first
      elsif i == (input.size - 1)
        position = :last
      end

      # apply rules to each character, breaking on the first one which returns output
      # for the word 'testify', the rules will see 'testify', 'estify', 'stify', 'tify', etc...
      subject = input[i..(input.size-1)]

      Cheferize.cheferize_rules.each do |rule|
        out = self.send(rule, subject, position)
        # uncomment to see to_chef in action.
        #puts "#{subject} #{' ('+position.to_s+') ' if position}: '#{rule.to_s}' returns '#{out.inspect}'"
        if out
          output += out[0]
          i += out[1]
          break
        end
      end
    end

    output
  end

  # Define a transformation rule.
  # Block must accept subject & position parameters.
  def self.cheferize_rule(name, &block)
    @rules ||= []
    sym = name.to_sym
    @rules << sym
    define_method(sym, block)
  end

  # Shows the transliteration rules.
  def self.cheferize_rules
    @rules
  end

  # General rules
  #
  # after a replacement occurs, skip to the next source character.  (only 1 transformation per letter)
  # if a multi-character test matches, skip over to the next chars.
  #
  # Rule API :
  #
  #  Each rule must accept parameters:
  #    - subject : a string to consider for transformation
  #    - position : a symbol indicating where the beginning of the subject string falls, in relationship
  #                 to the entire word being transformed.  Value will be :first, :last, or nil.
  #
  #  A rule may return either:
  #    - nil : Meaning this rule does not apply.
  #    - A array with element 0 containing a replacement string, and element 1 containing an integer describing the
  #      number of characters this match consumes in the entire word being transformed.
  #      Example : The string 'ew' is changed to 'oo' by the rule 'interior ew becomes oo'.  Because this match
  #                involves both the 'e' and the 'w', it consumes both characters.  The 'w' will not be considered
  #                for matching on its own.
  #
  #
  # Outline of rules.  Taken from encheferizer.php by Erik Bakker.
  #
  # for first char
  #
  #   ^e->i
  #   ^o->oo
  #
  # for non-first chars
  #
  #   ew->oo
  #   e$->e-a
  #   f->ff
  #   ir->ur, or i->ee if no i has occurred.
  #   ow->oo or o->u
  #   tion -> shun
  #   u->oo
  #
  #   An->Un
  #   Au->Oo
  #   A.->E (A not at the end becomes E)
  #   en$->ee
  #   e not at the beginning, do nothing. Pointless?
  #   th->tt
  #   the->zee
  #   v->f
  #   w->v
  #   pass

  # no recursive cheferizing yet.
  cheferize_rule "pass the bork" do |subject, position|
    [subject, 4] if position == :first && subject[0,4].downcase == 'bork'
  end

  cheferize_rule "randomly borkify after . or !" do |subject, position|
    if position == :last && (subject[0,1] == '.' || subject[0,1] == '!') && rand(3) == 0
      [', bork bork bork!', 1]
    end
  end

  # ^e->i
  cheferize_rule "initial e becomes i" do |subject, position|
    if position == :first && subject[0,1].downcase == 'e'
      [ChefString.new('i').in_same_case_as(subject[0,1]), 1]
    end
  end

  # ^o->oo
  cheferize_rule "initial o becomes oo" do |subject, position|
    if position == :first && subject[0,1].downcase == 'o'
      [ChefString.new('o').in_same_case_as(subject[0,1]) + 'o', 1]
    end
  end

  # ew->oo, if not at the beginning
  cheferize_rule "interior ew becomes oo" do |subject, position|
    ['oo', 2] if position != :first && subject[0,2] == 'ew'
  end

  # e$->e-a, if at the end
  cheferize_rule "final e becomes e-a" do |subject, position|
    ['e-a', 1] if position == :last && subject[0,1] == 'e'
  end

  # f->ff if not at the beginning
  cheferize_rule "interior f becomes ff" do |subject, position|
    ['ff', 1] if position != :first && subject[0,1] == 'f'
  end

  # ir->ur, or i->ee if no i has occurred.
  cheferize_rule "interior ir becomes ur, or first-occurring interior i becomes ee" do |subject, position|
    if position != :first
      if subject[0,2] == 'ir'
        ['ur', 2]
      elsif subject[0,1] == 'i' && !@chef_saw_an_i
        @chef_saw_an_i = true
        ['ee', 1]
      end
    end
  end

  # ow->oo or o->u
  cheferize_rule "ow becomes oo, or o becomes u" do |subject, position|
    if position != :first
      if subject[0,2] == 'ow'
        ['oo', 2]
      elsif subject[0,1] == 'o'
        ['u', 1]
      end
    end
  end

  # tion -> shun
  cheferize_rule "tion becomes shun" do |subject, position|
    if position != :first && subject[0,4] == 'tion'
      ['shun', 4]
    end
  end

  # u->oo
  cheferize_rule "interior u becomes oo" do |subject, position|
    if position != :first && subject[0,1] == 'u'
      ['oo', 1]
    end
  end

  # An->Un, an->un
  cheferize_rule "an becomes un" do |subject, position|
    if subject[0,2].downcase == 'an'
      [ChefString.new('u').in_same_case_as(subject[0,1]) + 'n', 2]
    end
  end

  # Au->Oo, au->oo
  cheferize_rule "au becomes oo" do |subject, position|
    if subject[0,2].downcase == 'au'
      [ChefString.new('o').in_same_case_as(subject[0,1]) + 'o', 2]
    end
  end

  # A.->E (A not at the end becomes E)
  cheferize_rule "non-final a becomes e" do |subject, position|
    if position != :last && subject[0,1].downcase == 'a'
      [ChefString.new('e').in_same_case_as(subject[0,1]), 1]
    end
  end

  # en$->ee
  cheferize_rule "final en becomes ee" do |subject, position|
    if position != :first && subject == 'en'
      ['ee', 2]
    end
  end

  # e not at the beginning, do nothing. Pointless?

  # the->zee
  cheferize_rule "the becomes zee" do |subject, position|
    if subject[0,3].downcase == 'the'
      [ChefString.new('z').in_same_case_as(subject[0,1]) + 'ee', 3]
    end
  end

  # th->t
  cheferize_rule "th becomes t" do |subject, position|
    if subject[0,2].downcase == 'th'
      [ChefString.new('t').in_same_case_as(subject[0,1]), 2]
    end
  end

  # v->f
  cheferize_rule "v becomes f" do |subject, position|
    if subject[0,1].downcase == 'v'
      [ChefString.new('f').in_same_case_as(subject[0,1]), 1]
    end
  end

  # w->v
  cheferize_rule "w becomes v" do |subject, position|
    if subject[0,1].downcase == 'w'
      [ChefString.new('v').in_same_case_as(subject[0,1]), 1]
    end
  end

  cheferize_rule "pass on anything else" do |subject, position|
    [subject[0,1], 1]
  end
end
